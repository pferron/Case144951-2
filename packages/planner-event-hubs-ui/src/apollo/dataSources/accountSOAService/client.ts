import { LoggerFactory } from '@cvent/logging/LoggerFactory';
import { AccountLocale, AccountVideoCenterConfig, AccountDetails } from '@cvent/planner-event-hubs-model/types';
import { CvestDataSource } from '@dataSources/CvestDataSource';
import { normalizeRawOrJSONResponseToJSON } from '@server/utils';

const LOG = LoggerFactory.create('account-soa-client');

export class AccountSOAClient extends CvestDataSource {
  constructor() {
    super();
    this.baseURL = `${process.env.ACCOUNT_SOA_SERVICE}/v1`;
  }

  /**
   * Fetches AccountConfig from Account SOA. Returns very small set
   * of AccountConfig data, just what's needed by VideoCenter today.
   *
   * @returns AccountConfig
   */
  public async getAccountConfig(): Promise<AccountVideoCenterConfig> {
    LOG.debug('getAccountConfig()');

    // RED
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const response = normalizeRawOrJSONResponseToJSON(await this.get('/accountconfig')) as any;

    return {
      AccountFeatures: {
        GeneralFeatures: {
          AIWritingAssistantEnabled: response?.AccountFeatures?.GeneralFeatures?.AIWritingAssistantEnabled,
          AllowCodeSnippets: response?.AccountFeatures?.GeneralFeatures?.AllowCodeSnippets,
          AllowCustomFonts: response?.AccountFeatures?.GeneralFeatures?.AllowCustomFonts,
          AllowGoogleAnalytics: response?.AccountFeatures?.GeneralFeatures?.AllowGoogleAnalytics,
          AllowOAuth: response?.AccountFeatures?.GeneralFeatures?.AllowOAuth
        },
        Blades: {
          AllowVideosBlade: response?.AccountFeatures?.Blades?.AllowVideosBlade
        }
      },
      VideoManagementFeatures: {
        VideoStorageSize: response?.VideoManagementFeatures?.VideoStorageSize,
        VideoCenterFlag: response?.VideoManagementFeatures?.VideoCenterFlag
      },
      InternationalSettings: {
        DefaultLanguageId: response?.AccountFeatures?.InternationalSettings?.DefaultLanguageId,
        DefaultCultureCode: response?.AccountFeatures?.InternationalSettings?.DefaultCultureCode
      },
      EventFeatures: {
        GeneralFeatures: {
          LicenseTypeId: response?.EventFeatures?.GeneralFeatures?.LicenseTypeId
        }
      }
    };
  }

  /**
   * Fetches AccountLocale from Account SOA.
   *
   * @returns AccountLocale
   */
  public async getAccountLocale(): Promise<AccountLocale[]> {
    LOG.debug('getAccountLocale()');

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return normalizeRawOrJSONResponseToJSON(await this.get('/accountlocale')) as any;
  }

  /**
   * Fetches AccountDetails from Account SOA.
   *
   * @returns AccountLocale
   */
  public async getAccountDetails(): Promise<AccountDetails> {
    LOG.debug('getAccountDetails()');

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return normalizeRawOrJSONResponseToJSON(await this.get('/accountdetailextended')) as any;
  }
}
