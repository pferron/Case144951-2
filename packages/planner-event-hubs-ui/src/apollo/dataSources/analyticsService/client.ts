import { CvestDataSource } from '@dataSources/CvestDataSource';
import { LoggerFactory } from '@cvent/auth-client';

const APPLICATION_NAME = 'Planner Event Hubs';

const LOG = LoggerFactory.create('analytics-api-client');
export interface AnalyticsField {
  fieldKey: string;
  fieldType: string;
  encrypt: boolean;
}

export interface AnalyticsRowData {
  values: Array<string>;
}

export interface VideoViewsAndPosition {
  videoId: string;
  totalViews: number;
  position: number;
}

export interface AnalyticsResponseData {
  total: number;
  fields: Array<AnalyticsField>;
  rows: Array<AnalyticsRowData>;
}

export class AnalyticsClient extends CvestDataSource {
  constructor() {
    super();
    this.baseURL = `${process.env.ANALYTICS_DATASOURCE_URL}`;
  }

  getTotalViewsByHubId = async (
    hubId: string,
    startDate: string,
    endDate: string,
    pipeline?: string
  ): Promise<AnalyticsResponseData> => {
    LOG.debug('getTotalViewsByHubId()', hubId);

    const params = [
      {
        name: 'startTime',
        value: startDate
      },
      {
        name: 'endTime',
        value: endDate
      }
    ];

    const requestBody = {
      context: {
        application: APPLICATION_NAME
      },
      criteria: {
        columns: ['time_window_per_day', 'entity_id', 'total_views'],
        expression: `entity_id = '${hubId}'`,
        params
      }
    };

    const dataSource = pipeline === 'v2' ? 'total_views_per_hub_per_day_agg_v2' : 'total_views_per_hub_per_day_agg';
    const data = (await this.post(`/v1/analytics_data_source/requestData/${dataSource}`, requestBody, {
      headers: { authorization: `API_KEY ${process.env.API_KEY}` }
    })) as AnalyticsResponseData;

    return data;
  };

  averageViewDurationByHubId = async (
    hubId: string,
    startDate: string,
    endDate: string,
    pipeline?: string
  ): Promise<AnalyticsResponseData> => {
    LOG.debug('averageViewDurationByHubId()', hubId);

    const params = [
      {
        name: 'startTime',
        value: startDate
      },
      {
        name: 'endTime',
        value: endDate
      }
    ];

    const requestBody = {
      context: {
        application: APPLICATION_NAME
      },
      criteria: {
        columns: ['time_window_per_day', 'entity_id', 'average_view_duration'],
        expression: `entity_id = '${hubId}'`,
        params
      }
    };

    const dataSource =
      pipeline === 'v2' ? 'average_view_duration_per_hub_per_day_agg_v2' : 'average_view_duration_per_hub_per_day_agg';
    const data = (await this.post(`/v1/analytics_data_source/requestData/${dataSource}`, requestBody, {
      headers: { authorization: `API_KEY ${process.env.API_KEY}` }
    })) as AnalyticsResponseData;

    return data;
  };

  getTotalVideosViewedByHubId = async (
    mappingId: string,
    hubId: string,
    startDate: string,
    endDate: string,
    pipeline?: string
  ): Promise<AnalyticsResponseData> => {
    LOG.debug('getTotalVideosViewedByHubId()', hubId, startDate, endDate);
    const params = [
      {
        name: 'mapping_id',
        value: mappingId
      },
      {
        name: 'startTime',
        value: startDate
      },
      {
        name: 'endTime',
        value: endDate
      }
    ];

    const requestBody = {
      context: {
        application: APPLICATION_NAME
      },
      criteria: {
        columns: ['time_window_per_day', 'video_id', 'entity_id', 'total_views'],
        expression: `entity_id = '${hubId}' AND view_type='Recording'`,
        params
      }
    };

    const dataSource =
      pipeline === 'v2' ? 'total_views_per_source_per_day_agg_v2' : 'total_views_per_source_per_day_agg';
    const data = await this.post(`/v1/analytics_data_source/requestData/${dataSource}`, requestBody, {
      headers: { authorization: `API_KEY ${process.env.API_KEY}` }
    });
    return data;
  };

  viewsByDeviceType = async (
    hubId: string,
    startDate: string,
    endDate: string,
    pipeline?: string
  ): Promise<AnalyticsResponseData> => {
    LOG.debug('viewsByDeviceType()', hubId);

    const params = [
      {
        name: 'startTime',
        value: startDate
      },
      {
        name: 'endTime',
        value: endDate
      }
    ];

    const requestBody = {
      context: {
        application: APPLICATION_NAME
      },
      criteria: {
        columns: ['time_window_per_day', 'entity_id', 'device_type', 'total_views'],
        expression: `entity_id = '${hubId}'`,
        params
      }
    };

    const dataSource =
      pipeline === 'v2'
        ? 'total_views_by_device_type_per_hub_per_day_agg_v2'
        : 'total_views_by_device_type_per_hub_per_day_agg';
    const data = (await this.post(`/v1/analytics_data_source/requestData/${dataSource}`, requestBody, {
      headers: { authorization: `API_KEY ${process.env.API_KEY}` }
    })) as AnalyticsResponseData;

    return data;
  };

  getAllVideosViewedByHubId = async (
    mappingId: string,
    hubId: string,
    startDate: string,
    endDate: string,
    limit: number,
    offset: number
  ): Promise<AnalyticsResponseData> => {
    LOG.debug('getAllVideosViewedByHubId()', hubId, startDate, endDate);
    const params = [
      {
        name: 'mapping_id',
        value: mappingId
      },
      {
        name: 'startTime',
        value: startDate
      },
      {
        name: 'endTime',
        value: endDate
      },
      {
        name: 'limit',
        value: limit
      },
      {
        name: 'offset',
        value: offset
      }
    ];

    const requestBody = {
      context: {
        application: APPLICATION_NAME
      },
      criteria: {
        columns: ['video_id', 'total_views'],
        expression: `entity_id = '${hubId}'`,
        params
      }
    };

    return this.post(`/v1/analytics_data_source/requestData/total_views_per_hub_per_video_agg_v2`, requestBody, {
      headers: { authorization: `API_KEY ${process.env.API_KEY}` }
    });
  };

  getMemberVideoWatchDurationByHubId = async (
    mappingId: string,
    hubId: string,
    videoId: string,
    startDate: string,
    endDate: string,
    limit: number,
    offset: number
  ): Promise<AnalyticsResponseData> => {
    LOG.debug('getMemberVideoWatchDurationByHubId()', hubId, videoId, startDate, endDate);
    const params = [
      {
        name: 'mapping_id',
        value: mappingId
      },
      {
        name: 'startTime',
        value: startDate
      },
      {
        name: 'endTime',
        value: endDate
      },
      {
        name: 'limit',
        value: limit
      },
      {
        name: 'offset',
        value: offset
      }
    ];

    const requestBody = {
      context: {
        application: APPLICATION_NAME
      },
      criteria: {
        columns: ['unique_id', 'total_duration'],
        expression: `entity_id = '${hubId}' AND video_id = '${videoId}'`,
        params
      }
    };

    return this.post(`/v1/analytics_data_source/requestData/total_watch_duration_per_member_agg_v2`, requestBody, {
      headers: { authorization: `API_KEY ${process.env.API_KEY}` }
    });
  };
}
