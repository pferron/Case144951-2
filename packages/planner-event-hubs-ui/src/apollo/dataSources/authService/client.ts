import { LoggerFactory } from '@cvent/logging/LoggerFactory';
import { GrantedAccessToken } from '@cvent/auth-client';
import { AUTH_COOKIE_MAX_AGE } from '@utils/constants';
import { RequestOptions, RESTDataSource } from '../RESTDataSource/RESTDataSource';

const LOG = LoggerFactory.create('auth-api');

/**
 * This class does not extend CvestDataSource because we don't want to set all the same headers and params from here.
 * The intent of this class is to provide a path to generate a bearer for public/anonymous users when visiting a public Video Center.
 *
 * This class also provides a clear and concise example of a DataSource that does not use CvestDataSource.
 *
 * If further work is needed here, consider removing this class and using @cvent/auth-client instead.
 */
export class AuthClient extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = `${process.env.AUTH_SERVICE}/v1`;
  }

  public willSendRequest(request: RequestOptions): void {
    this.setStandardHeaders(request);
  }

  public setStandardHeaders(request: RequestOptions): void {
    request.headers.set('Authorization', `API_KEY ${process.env.API_KEY}`);
  }

  // TODO: remove publicLogin from authService
  public async publicLogin(
    hubId: string,
    environment: string,
    accountMappingId: string,
    accountId: string
  ): Promise<GrantedAccessToken> {
    LOG.debug('publicLogin sending metadata', { accountMappingId, hubId, environment, accountId });
    return this.post(`/access_token?accessTokenTTL=${AUTH_COOKIE_MAX_AGE}`, {
      metadata: {
        accountMappingId,
        VideoCenterId: hubId,
        environment,
        accountId,
        AccountId: accountId
      },
      grantedAuthorizations: [
        {
          appId: process.env.VIDEO_HUB_SERVICE_APP_ID,
          roles: ['video-center:anonymous']
        },
        {
          appId: process.env.UNIVERSAL_VIDEO_APP_ID,
          roles: ['videos:read']
        },
        {
          appId: process.env.UNIVERSAL_CONTACTS_APP_ID,
          roles: ['contacts:read']
        }
      ]
    });
  }

  public async verifyToken(accessToken: string): Promise<GrantedAccessToken> {
    LOG.debug('verifying Token');
    return this.get(`/access_token/${accessToken}/verify`);
  }

  public async accessTokenRefresh(accessToken: string): Promise<void> {
    LOG.debug('accessToken Refreshing');
    return this.put(`/access_token/${accessToken}/refresh`);
  }
}
