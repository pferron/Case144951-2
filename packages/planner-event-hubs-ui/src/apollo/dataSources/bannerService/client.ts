import { LoggerFactory } from '@cvent/logging/LoggerFactory';
import { CvestDataSource } from '@dataSources/CvestDataSource';
import {
  NewBanner,
  BannerPagingResponse,
  ExistingBanner,
  FilterInput,
  BannerUpdate
} from '@cvent/planner-event-hubs-model/types';

const LOG = LoggerFactory.create('banner-service-client');
export class BannerServiceClient extends CvestDataSource {
  constructor() {
    super();
    this.baseURL = `${process.env.VIDEO_HUB_SERVICE}/v1/`;
  }

  getBanner = async (centerId: string, bannerId: string): Promise<ExistingBanner> => {
    LOG.debug('getBanner(bannerId)', bannerId);
    return this.get(`video-hubs/${centerId}/banners/${bannerId}`);
  };

  updateBanner = async (bannerUpdate: BannerUpdate): Promise<ExistingBanner> => {
    LOG.debug('updateBanner(bannerId)', bannerUpdate);
    return this.put(`video-hubs/${bannerUpdate.centerId}/banners/${bannerUpdate.id}`, bannerUpdate, {
      cacheRefreshOptions: { update: true }
    });
  };

  getBanners = async (centerId: string, filterInput: FilterInput): Promise<BannerPagingResponse> => {
    LOG.debug('getBanners(filterInput)', filterInput);
    const params: { token?: string; limit?: number } = {};

    if (filterInput?.token) {
      params.token = filterInput.token;
    }

    if (filterInput?.limit) {
      params.limit = filterInput.limit;
    }
    return this.get(`video-hubs/${centerId}/banners`, params);
  };

  bannerCreate = async (banner: NewBanner): Promise<ExistingBanner> => {
    LOG.debug('bannerCreate(banner)', banner);
    return this.post(`video-hubs/${banner.centerId}/banners`, banner);
  };

  bannerDelete = async (centerId: string, bannerId: string): Promise<void> => {
    LOG.debug('bannerDelete(bannerId)', bannerId);
    return this.delete(`video-hubs/${centerId}/banners/${bannerId}`, undefined, {
      cacheRefreshOptions: { evict: true }
    });
  };

  getBannersByIds = async (centerId: string, bannerIds: string[]): Promise<ExistingBanner[]> => {
    LOG.debug(`getBannersByIds (${bannerIds.length})`);

    return Promise.all(
      bannerIds.map(async (id): Promise<ExistingBanner> => {
        return this.getBanner(centerId, id);
      })
    );
  };
}
