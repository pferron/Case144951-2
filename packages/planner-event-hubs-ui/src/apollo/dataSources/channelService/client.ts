import { CvestDataSource } from '@dataSources/CvestDataSource';
import { ChannelBannerInput, ChannelInput, ChannelStatus, ImageInput } from '@cvent/planner-event-hubs-model/types';
import { getRequestOptionsWithCacheOptions } from '@utils/util';
import { NOT_FOUND_ERROR_CODE } from '@utils/constants';

export interface ChannelResponse {
  id: string;
  status: ChannelStatus;
  catalogId: string;
  title: string;
  description: string;
  imageUrl: string;
  banners: string[];
}

export interface ImageResponse {
  id: string;
  relativePath: string;
  optimizedUrl: string;
  createdAt: string;
  filename: string;
  size: number;
}

export class ChannelServiceClient extends CvestDataSource {
  constructor() {
    super();
    this.baseURL = `${process.env.VIDEO_HUB_SERVICE}/v1/`;
  }

  getChannel = async (channelId: string): Promise<ChannelResponse> => {
    return this.get(`channels/${channelId}`, {}, getRequestOptionsWithCacheOptions(false, {}));
  };

  getChannelImage = async (channelId: string): Promise<ImageResponse> => {
    // We don't want to throw error if no image is added to a channel
    try {
      return await this.get(`channels/${channelId}/images`, {}, getRequestOptionsWithCacheOptions(false, {}));
    } catch (e) {
      if (e.code === NOT_FOUND_ERROR_CODE) {
        return null;
      }
      throw e;
    }
  };

  createChannel = async (title: string, description: string): Promise<ChannelResponse> => {
    return this.post('channels', { title, description });
  };

  deleteChannelImage = async (channelId: string, imageId: string): Promise<void> => {
    return this.delete(`channels/${channelId}/images/${imageId}`);
  };

  associateChannelHub = async (hubId: string, channelId: string, shortUrl: string): Promise<void> => {
    return this.post(`channels/${channelId}/video-hubs/${hubId}`, { shortUrl });
  };

  deleteChannel = async (channelId: string): Promise<void> => {
    return this.delete(`channels/${channelId}`);
  };

  uploadChannelImage = async (channelId: string, image: ImageInput): Promise<ImageResponse> => {
    return this.post(`channels/${channelId}/images`, image);
  };

  updateChannel = async (channelInput: ChannelInput): Promise<ChannelResponse> => {
    return this.put(`channels/${channelInput.id}`, channelInput);
  };

  associateChannelCatalog = async (channelId: string, catalogId: string): Promise<void> => {
    return this.post(`channels/${channelId}/catalogs/${catalogId}`);
  };

  disassociateChannelCatalog = async (channelId: string, catalogId: string): Promise<void> => {
    return this.delete(`channels/${channelId}/catalogs/${catalogId}`);
  };

  createChannelBannerAssociation = async (channelBannerInput: ChannelBannerInput): Promise<void> => {
    return this.put(`channels/${channelBannerInput.channel}/banners/${channelBannerInput.banner}`, {
      channel: { id: channelBannerInput.channel },
      banner: { id: channelBannerInput.banner },
      order: channelBannerInput.order
    });
  };

  deleteChannelBannerAssociation = async (channelBannerInput: ChannelBannerInput): Promise<void> => {
    return this.delete(`channels/${channelBannerInput.channel}/banners/${channelBannerInput.banner}`);
  };
}
