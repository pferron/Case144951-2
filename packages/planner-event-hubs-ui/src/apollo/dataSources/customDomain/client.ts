import { CvestDataSource } from '@dataSources/CvestDataSource';
import { getRequestOptionsWithCacheOptions } from '@utils/util';
import { CustomDomain, CustomDomainMapping } from '@cvent/planner-event-hubs-model/types';
import { getAccountId } from '@resolvers/common/utils/authMetadataUtils';
import { NOT_FOUND_ERROR_CODE } from '@utils/constants';

export class CustomDomainServiceClient extends CvestDataSource {
  constructor() {
    super();
    this.baseURL = `${process.env.CUSTOM_DOMAIN_SERVICE_URL}/v1`;
  }

  getCustomDomainForHub = async (hubId: string): Promise<CustomDomainMapping> => {
    try {
      const response = await this.get(
        `/custom-domains/events-plus/${hubId}`,
        {},
        getRequestOptionsWithCacheOptions(true)
      );
      return {
        customDomainId: response.accountDomain.id,
        entityId: response.id,
        trailingName: response.customUrlName
      };
    } catch (error) {
      if (error.code === NOT_FOUND_ERROR_CODE) {
        return null;
      }
      throw error;
    }
  };

  getCustomDomainForAccount = async (): Promise<Array<CustomDomain>> => {
    const accountId = getAccountId(this.context.auth);
    return this.get(
      `/custom-domain/${accountId}`,
      {},
      getRequestOptionsWithCacheOptions(true, { headers: { authorization: `API_KEY ${process.env.API_KEY}` } })
    );
  };

  createCustomDomainMapping = async (hubCustomDomain: CustomDomainMapping): Promise<CustomDomainMapping> => {
    const response = await this.post(
      `/custom-domains/events-plus/${hubCustomDomain.entityId}`,
      {
        accountDomain: { id: hubCustomDomain.customDomainId },
        landingPageId: 0,
        customUrlName: hubCustomDomain.trailingName
      },
      getRequestOptionsWithCacheOptions(false, {})
    );
    return {
      customDomainId: response.accountDomain.id,
      entityId: response.id,
      trailingName: response.customUrlName
    };
  };

  updateCustomDomainMapping = async (hubCustomDomain: CustomDomainMapping): Promise<CustomDomainMapping> => {
    const response = await this.put(
      `/custom-domains/events-plus/${hubCustomDomain.entityId}`,
      {
        accountDomain: { id: hubCustomDomain.customDomainId },
        landingPageId: 0,
        customUrlName: hubCustomDomain.trailingName
      },
      getRequestOptionsWithCacheOptions(false, {})
    );
    return {
      customDomainId: response.accountDomain.id,
      entityId: response.id,
      trailingName: response.customUrlName
    };
  };

  deleteCustomDomainMapping = async (hubId: string): Promise<boolean> => {
    await this.delete(`/custom-domains/events-plus/${hubId}`);
    return true;
  };
}
