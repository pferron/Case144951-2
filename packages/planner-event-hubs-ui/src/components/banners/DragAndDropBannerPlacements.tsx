import { useTable } from '@cvent/carina/components/Table';
import { DraggableTableProps } from '@cvent/carina/components/Table/DraggableTable';
import dynamic from 'next/dynamic';
import { useContext, useEffect, useMemo, useState } from 'react';

import TextLink from '@cvent/carina/components/TextLink';
import { ThemeContext } from '@cvent/carina/components/ThemeProvider/ThemeContext';
import { injectTestId } from '@cvent/nucleus-test-automation';
import { useTranslate } from 'nucleus-text';

import ContentCard from '@components/ContentCard';
import { BANNERS_PATH_PARAM_KEY, BANNER_URL, VIDEO_HUB_PATH_PARAM, VIDEO_HUB_PATH_PARAM_KEY } from '@utils/constants';
import { useRouter } from 'next/router';

import { useMutation, useQuery } from '@apollo/client';
import { ExistingBannerAssociationWithBanner } from '@cvent/planner-event-hubs-model/types';
import useQueryParams from '@hooks/useQueryParam';
import { UPDATE_BANNER_ASSOCIATIONS } from '@cvent/planner-event-hubs-model/operations/banner';
import { getChannelQuery } from '@cvent/planner-event-hubs-model/operations/channel';

const MAX_PLACEMENTS = 6;

const DraggableTable = dynamic<DraggableTableProps>(() =>
  import('@cvent/carina/components/Table').then(mod => mod.DraggableTable)
);

const bannerAssociationUpdate = fields =>
  fields.map((f, index) => ({
    banner: f.banner.id,
    displayOrder: index + 1
  }));

const useStyles = () => {
  const appTheme = useContext(ThemeContext);

  return {
    row: {
      fontSize: appTheme.font.base.size.m,
      display: 'flex',
      gap: 5,
      justifyContent: 'space-between',
      '& div': {
        fontSize: appTheme.font.base.size.m
      }
    },
    description: {
      marginTop: 8,
      '& div': {
        fontSize: appTheme.font.base.size.xs,
        color: appTheme.font.color.soft
      }
    },
    footer: {
      marginTop: 8,
      fontSize: appTheme.font.base.size.s,
      color: appTheme.font.color.soft
    }
  };
};

interface Props {
  totalBannersNumber?: number;
  data: Array<ExistingBannerAssociationWithBanner>;
}

function DragAndDropBannerPlacements({ data }: Props): JSX.Element {
  const { row, description, footer } = useStyles();
  const { translate } = useTranslate();
  const router = useRouter();
  const query = useQueryParams();
  const centerId = query[VIDEO_HUB_PATH_PARAM] as string;

  const [updateBannerAssociationMutation] = useMutation(UPDATE_BANNER_ASSOCIATIONS);

  const [bannerPlacements, setPlacementList] = useState([]);

  const tableRows = useMemo(() => bannerPlacements, [bannerPlacements]);

  const { data: channelData } = useQuery(getChannelQuery, {
    variables: {
      channelId: data[0].entityId
    },
    skip: !(data[0].entityType === 'Channel' && data[0].entityId)
  });

  useEffect(() => {
    setPlacementList(() =>
      data.map((item, index: number) => ({
        ...item,
        rowName: `${index}`
      }))
    );
  }, [data]);

  const onUpdateBannerAssociations = async (updatedAssociations): Promise<void> => {
    const associationsToSave = bannerAssociationUpdate(updatedAssociations);
    const associatedCenterId = updatedAssociations[0].centerId;
    const associatedEntityId = updatedAssociations[0].entityId;
    const associatedEntityType = updatedAssociations[0].entityType;
    await updateBannerAssociationMutation({
      variables: {
        input: {
          centerId: associatedCenterId.toString(),
          entityId: associatedEntityId.toString(),
          entityType: associatedEntityType.toString(),
          bannerAssociation: associationsToSave
        }
      }
    });
  };
  const [Table, TableColumn] = useTable({ data: tableRows, isDraggable: true, DraggableTable });

  const onDragEndCallback = ({ fromIndex, toIndex }: { fromIndex?: number; toIndex?: number }) => {
    const newPlacementList = [...bannerPlacements];
    if (fromIndex !== undefined && toIndex !== undefined) {
      newPlacementList.splice(toIndex, 0, newPlacementList.splice(fromIndex, 1)[0]);
      setPlacementList(newPlacementList);
      onUpdateBannerAssociations(newPlacementList);
    }
  };

  const onEdit = (bannerId: string): void => {
    if (bannerId) {
      const redirection = BANNER_URL.replace(VIDEO_HUB_PATH_PARAM_KEY, centerId).replace(
        BANNERS_PATH_PARAM_KEY,
        bannerId
      );
      router.push({
        pathname: redirection
      });
    }
  };

  const placementCellRenderer = (_, { rowData }): JSX.Element => (
    <div {...injectTestId('banner-placement-table-row')} css={row}>
      <div>
        <div {...injectTestId(`banner-placement-table-row-${rowData.rowName}`)}>{rowData?.banner?.name}</div>
        <div css={description} {...injectTestId(`banner-placement-table-row-visibility-${rowData.rowName}`)} />
      </div>
      <TextLink element="button" size="l" onClick={() => onEdit(rowData.banner.id)}>
        {translate('BannerPlacements-Edit-Banner-Link')}
      </TextLink>
    </div>
  );

  const pageName = data[0].entityType !== 'Channel' ? 'Homepage' : channelData?.getChannelInformation?.title;

  const dragAndDropBannerPlacementHeader: JSX.Element = <h3 style={{ margin: 0 }}>{pageName}</h3>;

  const contentCardBodyContent: JSX.Element = (
    <>
      <Table allowDragAndDrop rowMode="drag-and-drop" onDragEnd={onDragEndCallback} hideHeader paddingMode="roomy">
        <TableColumn name="name" cellRenderer={placementCellRenderer} />
      </Table>
      <div css={footer}>
        {translate('BannerPlacements-NumBannersAssigned-Footer', {
          numBannersAssigned: data.length.toString(),
          totalNumBanners: MAX_PLACEMENTS
        })}
      </div>
    </>
  );

  return (
    <ContentCard
      collapsible
      containerStyle={{ margin: 24 }}
      cardName={data[0].entityType}
      header={dragAndDropBannerPlacementHeader}
      BodyComponent={contentCardBodyContent}
    />
  );
}

export default DragAndDropBannerPlacements;
