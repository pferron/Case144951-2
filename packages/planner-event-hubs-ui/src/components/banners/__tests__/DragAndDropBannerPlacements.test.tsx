import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { TestWrapper } from '@utils/testUtil/TestWrapper';
import { MockedProvider as ApolloProvider } from '@apollo/client/testing';
import 'jest-axe/extend-expect';
import { axe } from 'jest-axe';
import DragAndDropBannerPlacements from '../DragAndDropBannerPlacements';
import bannersAssociationList from '../fixtures/bannersAssociationList.json';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: mockHistoryPush
  })
}));

describe('DragAndDropBannerPlacements', () => {
  it('should render correct elements on page with no violations', async () => {
    const { container } = render(
      <ApolloProvider>
        <TestWrapper>
          <DragAndDropBannerPlacements
            data={bannersAssociationList.data.bannerAssociations.data}
            totalBannersNumber={6}
          />
        </TestWrapper>
      </ApolloProvider>
    );
    expect(container).toMatchSnapshot();
    await waitFor(() => expect(screen.getAllByTestId('banner-placement-table-row').length).toEqual(4));
    expect(await axe(container)).toHaveNoViolations();
  });
  it('should render banner placements in correct order', async () => {
    const { container } = render(
      <ApolloProvider>
        <TestWrapper>
          <DragAndDropBannerPlacements
            data={bannersAssociationList.data.bannerAssociations.data}
            totalBannersNumber={6}
          />
        </TestWrapper>
      </ApolloProvider>
    );
    expect(screen.getByTestId('banner-placement-table-row-0')).toHaveTextContent(/Gogogo new banner/);
    expect(await axe(container)).toHaveNoViolations();
  });

  it('should have edit link on the banner item', async () => {
    const { container } = render(
      <ApolloProvider>
        <TestWrapper>
          <DragAndDropBannerPlacements
            data={bannersAssociationList.data.bannerAssociations.data}
            totalBannersNumber={6}
          />
        </TestWrapper>
      </ApolloProvider>
    );
    const editLinks = await screen.findAllByText('Edit'); // findAllByText returns an array
    expect(editLinks.length > 0).toBeTruthy();
    const editLink = editLinks[0];
    expect(editLink).toBeInTheDocument();
    fireEvent.click(editLink);
    expect(await axe(container)).toHaveNoViolations();
  });
});
