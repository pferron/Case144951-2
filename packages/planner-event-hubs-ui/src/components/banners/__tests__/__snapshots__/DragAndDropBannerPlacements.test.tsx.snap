// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragAndDropBannerPlacements should render correct elements on page with no violations 1`] = `
.emotion-0 {
  padding: 24px;
  margin: 0;
  border-radius: 4px;
  background-color: rgb(255,255,255);
  margin: 24px;
}

.emotion-1 {
  font-family: 'Rubik','Helvetica','Arial';
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  box-sizing: border-box;
  padding-bottom: 24px;
}

.emotion-1 h3 {
  margin-bottom: 20px;
}

.emotion-1 p {
  font-family: 'Rubik','Helvetica','Arial';
}

.emotion-2 {
  width: 100%;
}

.emotion-3 {
  background: transparent;
  border: none;
  -webkit-transform: rotate(-180deg);
  -moz-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  padding: 0;
  -webkit-transition: -webkit-transform .3s ease-in-out;
  transition: transform .3s ease-in-out;
}

.emotion-4 {
  width: 1.5rem;
  height: 1.5rem;
  fill: rgb(26,32,38);
}

.emotion-5 {
  max-height: 3000px;
  -webkit-transition: max-height .3s ease-in-out;
  transition: max-height .3s ease-in-out;
  box-sizing: border-box;
}

.emotion-5 h3 {
  font-family: 'Rubik','Helvetica','Arial';
  margin-bottom: 16px;
}

.emotion-5 h5 {
  font-family: 'Rubik','Helvetica','Arial';
  text-transform: uppercase;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  margin-bottom: 16px;
}

.emotion-5 p {
  margin-bottom: 16px;
  font-size: 0.875rem;
}

.emotion-5 a {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #006ae1;
}

.emotion-6 {
  border-top: 0;
  padding-top: 0;
}

.emotion-8 {
  margin-top: 8px;
  font-size: 0.875rem;
  color: rgb(105,113,122);
}

<div>
  <div
    class="emotion-0"
    data-cvent-id="Homepage-section"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <h3
          style="margin: 0px;"
        >
          Homepage
        </h3>
      </div>
      <button
        aria-controls="container-body_key_accordion"
        aria-expanded="true"
        aria-label="Collapse"
        class="emotion-3"
        data-cvent-id="Homepage-toggle-card-collapse"
        id="accordion_key_accordion"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="emotion-4"
          focusable="false"
          role="img"
          viewBox="0 0 32 32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <svg
              viewBox="0 0 24 24"
            >
              <path
                d="M18.7 8.3c-.4-.4-1-.4-1.4 0L12 13.6 6.7 8.3c-.4-.4-1-.4-1.4 0s-.4 1 0 1.4l6 6c.2.2.5.3.7.3s.5-.1.7-.3l6-6c.4-.4.4-1 0-1.4z"
              />
            </svg>
          </g>
        </svg>
      </button>
    </div>
    <div
      aria-labelledby="accordion_key_accordion"
      class="emotion-5"
      id="container-body_key_accordion"
      role="region"
    >
      <div
        class="emotion-6"
      >
        <div
          class="emotion-7"
        >
          <div
            class="emotion-8"
          >
            4/6 banners
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
